File: tools/test_news_portal_renderer.py
"""Focused static contract for article ID 53 news portal renderer."""
from pathlib import Path

SOURCE = Path(__file__).parent.parent / "templates/pn_natuna_2026/html/com_content/article/default.php"
CSS = Path(__file__).parent.parent / "templates/pn_natuna_2026/css/template.css"
text = SOURCE.read_text(encoding="utf-8")
css = CSS.read_text(encoding="utf-8")

# Fixture edge cases required by editorial-card normalization contract.
fixtures = [
    ({"image_fulltext": "", "image_intro": "images/news/intro.webp"}, "/images/news/intro.webp"),
    ({"image_fulltext": "images/news/full.webp", "image_intro": "images/news/intro.webp"}, "/images/news/full.webp"),
]
for images, expected in fixtures:
    selected = (images.get("image_fulltext") or "").strip() or (images.get("image_intro") or "").strip()
    normalized = "/" + selected.lstrip("/") if selected and not selected.startswith(("http://", "https://", "//")) else selected
    assert normalized == expected

# Sentinel publish_up must use created date.
publish_up, created = "0000-00-00 00:00:00", "2026-07-11 08:30:00"
assert (publish_up if publish_up > "2000-01-02 00:00:00" else created) == created

for requirement in [
    "(int) $item->id === 53",
    "portalNews = $portalItems(12, 3, 'portalNewsCategory')",
    "portalAnnouncements = $portalItems(13, 5, 'portalAnnouncementCategory')",
    "NOT LIKE ' . $db->quote('berita-dan-pengumuman%')",
    "trim((string) ($decoded['image_fulltext'] ?? '')) ?: trim((string) ($decoded['image_intro'] ?? ''))",
    "'a.images', 'a.introtext', 'a.fulltext'",
    "publishUp > '2000-01-02 00:00:00' ? $publishUp : $created",
    "/images/hero/gedung-pn-natuna-2026.webp",
    "news-portal__news-card",
    "news-portal__announcements",
    "RouteHelper::getArticleRoute",
]:
    assert requirement in text, requirement

for requirement in [
    "NEWS PORTAL 2026-07-11",
    "height: clamp(210px, 58vw, 240px)",
    "grid-template-columns: repeat(3, minmax(0, 1fr))",
    "body.is-dark .news-portal",
    "prefers-reduced-motion: reduce",
]:
    assert requirement in css, requirement

print("news portal renderer contract: ok")

File: tools/test_news_portal_renderer.php
<?php
/** Focused contract check for article ID 53 portal renderer. */
$source = (string) file_get_contents(__DIR__ . '/../templates/pn_natuna_2026/html/com_content/article/default.php');
$failures = [];
$expect = static function (bool $condition, string $message) use (&$failures): void {
    if (!$condition) {
        $failures[] = $message;
    }
};

$expect(str_contains($source, '(int) $item->id === 53'), 'Missing article ID 53 portal branch.');
$expect(str_contains($source, "trim((string) (\$images['image_fulltext'] ?? '')) ?: trim((string) (\$images['image_intro'] ?? ''))"), 'Images must prefer non-empty fulltext then intro.');
$expect(str_contains($source, "'a.images', 'a.introtext', 'a.fulltext'"), 'Portal query must fetch article bodies for automatic first-image fallback.');
$expect(str_contains($source, "(string) (\$portalItem->introtext ?? '')") && str_contains($source, "(string) (\$portalItem->fulltext ?? '')"), 'Portal cards must tolerate partial article objects without emitting undefined-property warnings.');
$expect(str_contains($source, "'2000-01-02 00:00:00'"), 'Dates must fall back from sentinel publish_up to created.');
$expect(str_contains($source, 'portalNews = $portalItems(12, 3, \'portalNewsCategory\')'), 'Missing category 12 query.');
$expect(str_contains($source, 'portalAnnouncements = $portalItems(13, 5, \'portalAnnouncementCategory\')'), 'Missing category 13 query.');
$expect(str_contains($source, 'news-portal__hero'), 'Missing semantic news portal hero.');
$expect(str_contains($source, '/images/hero/gedung-pn-natuna-2026.webp'), 'Portal must use courthouse hero WebP.');
$expect(str_contains($source, 'news-portal__news-card'), 'Missing news cards.');
$expect(str_contains($source, 'news-portal__announcement'), 'Missing announcement rows.');
$expect(str_contains($source, 'RouteHelper::getArticleRoute'), 'Portal links must use Joomla routes.');

if ($failures) {
    fwrite(STDERR, implode(PHP_EOL, $failures) . PHP_EOL);
    exit(1);
}
echo "news portal renderer contract: ok\n";
